src/rgw/rgw_process.cc

```diff
@@ -87,12 +90,79 @@ void RGWProcess::RGWWQ::_process(RGWRequest *req, ThreadPool::TPHandle &) {
   process->req_throttle.put(1);
   perfcounter->inc(l_rgw_qactive, -1);
 }
+bool rate_limit(rgw::sal::Store* store, req_state* s) {
+  // we dont want to limit health check or system or admin requests
+  const auto& is_admin_or_system = s->user->get_info();
+  if (s->op_type ==  RGW_OP_GET_HEALTH_CHECK || is_admin_or_system.admin || is_admin_or_system.system)
+    return false;
+  std::string userfind;
+  RGWQoSInfo global_user;
+  RGWQoSInfo global_bucket;
+  RGWQoSInfo global_anon;
+  RGWQoSInfo* bucket_qos;
+  RGWQoSInfo* user_qos;
+  store->get_qos(global_bucket, global_user, global_anon);
+  bucket_qos = &global_bucket;
+  user_qos = &global_user;
+  s->user->get_id().to_str(userfind);
+  userfind = "u" + userfind;
+  s->ratelimit_user_name = userfind;
+  std::string bucketfind = !rgw::sal::Bucket::empty(s->bucket.get()) ? "b" + s->bucket->get_marker() : "";
+  s->ratelimit_bucket_name = bucketfind;
+  const char *method = s->info.method;
+
+  auto iter = s->user->get_attrs().find(RGW_ATTR_RATELIMIT);
+  if(iter != s->user->get_attrs().end()) {
+    try {
+      RGWQoSInfo user_qos_temp;
+      bufferlist& bl = iter->second;
+      auto biter = bl.cbegin();
+      decode(user_qos_temp, biter);
+      // override global rate limiting only if local rate limiting is enabled
+      if (user_qos_temp.enabled)
+        *user_qos = user_qos_temp;
+    } catch (buffer::error& err) {
+      dout(0) << "ERROR: failed to decode rate limit" << dendl;
+      return -EIO;
+    }
+  }
+  if(!rgw::sal::Bucket::empty(s->bucket.get()))
+  {
+    iter = s->bucket->get_attrs().find(RGW_ATTR_RATELIMIT);
+    if(iter != s->bucket->get_attrs().end()) {
+      try {
+        RGWQoSInfo bucket_qos_temp;
+        bufferlist& bl = iter->second;
+        auto biter = bl.cbegin();
+        decode(bucket_qos_temp, biter);
+        // override global rate limiting only if local rate limiting is enabled
+        if (bucket_qos_temp.enabled)
+          *bucket_qos = bucket_qos_temp;
+      } catch (buffer::error& err) {
+        dout(0) << "ERROR: failed to decode rate limit" << dendl;
+        return -EIO;
+      }
+    }
+  }
+
+  if (s->user->get_id().id == RGW_USER_ANON_ID && global_anon.enabled) {
+      *user_qos = global_anon;
+  }
+  bool limit_bucket = false;
+  bool limit_user = s->qos_data->should_rate_limit(method, s->ratelimit_user_name, s->time, user_qos, s);
+  if(!rgw::sal::Bucket::empty(s->bucket.get()) && !limit_user)
+    limit_bucket = s->qos_data->should_rate_limit(method, s->ratelimit_bucket_name, s->time, bucket_qos, s);
```

Comment:
this would be a good place to log why a request is being rejected. for example, something like
> bucket rate limit exceeded (ops=X bw=Y)